src/shell/shell_cmds.cpp

```diff
@@ -465,26 +472,250 @@ static std::string to_search_pattern(const char *arg)
+		return true;
+	}
+	if (!optB&&!optS) {
+		shell->WriteOut(MSG_Get("SHELL_CMD_DIR_INTRO"),uselfn&&!optZ&&DOS_GetSFNPath(path,largs,true)?largs:sargs);
+		if (optP) {
+			p_count+=optW?5:1;
+			if (p_count%(GetPauseCount()*w_size)<2) {
+				shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));
+				Bit8u c;Bit16u n=1;
+				DOS_ReadFile(STDIN,&c,&n);
+				if (c==3) {shell->WriteOut("^C\r\n");return false;}
+				if (c==0) DOS_ReadFile(STDIN,&c,&n); // read extended key
+				shell->WriteOut_NoParsing("\n");
+			}
+		}
+	}
+	if (*(sargs+strlen(sargs)-1) != '\\') safe_strcat(sargs,"\\");
+
+	Bit32u cbyte_count=0,cfile_count=0,w_count=0;
+	int fbak=lfn_filefind_handle;
+	lfn_filefind_handle=uselfn&&!optZ?LFN_FILEFIND_INTERNAL:LFN_FILEFIND_NONE;
+	bool ret=DOS_FindFirst(args,0xffff & ~DOS_ATTR_VOLUME), found=true, first=true;
+	lfn_filefind_handle=fbak;
+	if (ret) {
+		std::vector<DtaResult> results;
+
+		lfn_filefind_handle=uselfn&&!optZ?LFN_FILEFIND_INTERNAL:LFN_FILEFIND_NONE;
+		do {    /* File name and extension */
+			DtaResult result;
+			dta.GetResult(result.name,result.size,result.date,result.time,result.attr);
+			safe_strcpy(result.lname,result.name);
+
+			/* Skip non-directories if option AD is present, or skip dirs in case of A-D */
+			if(optAD && !(result.attr&DOS_ATTR_DIRECTORY) ) continue;
+			else if(optAminusD && (result.attr&DOS_ATTR_DIRECTORY) ) continue;
+			else if(optAS && !(result.attr&DOS_ATTR_SYSTEM) ) continue;
+			else if(optAminusS && (result.attr&DOS_ATTR_SYSTEM) ) continue;
+			else if(optAH && !(result.attr&DOS_ATTR_HIDDEN) ) continue;
+			else if(optAminusH && (result.attr&DOS_ATTR_HIDDEN) ) continue;
+			else if(optAR && !(result.attr&DOS_ATTR_READ_ONLY) ) continue;
+			else if(optAminusR && (result.attr&DOS_ATTR_READ_ONLY) ) continue;
+			else if(optAA && !(result.attr&DOS_ATTR_ARCHIVE) ) continue;
+			else if(optAminusA && (result.attr&DOS_ATTR_ARCHIVE) ) continue;
+			else if(!(optA||optAD||optAminusD||optAS||optAminusS||optAH||optAminusH||optAR||optAminusR||optAA||optAminusA) && (result.attr&DOS_ATTR_SYSTEM || result.attr&DOS_ATTR_HIDDEN) && strcmp(result.name, "..") ) continue;
+
+			results.push_back(result);
+
+		} while ( (ret=DOS_FindNext()) );
+		lfn_filefind_handle=fbak;
+
+		if (optON) {
+			// Sort by name
+			std::sort(results.begin(), results.end(), DtaResult::compareName);
+		} else if (optOE) {
+			// Sort by extension
+			std::sort(results.begin(), results.end(), DtaResult::compareExt);
+		} else if (optOD) {
+			// Sort by date
+			std::sort(results.begin(), results.end(), DtaResult::compareDate);
+		} else if (optOS) {
+			// Sort by size
+			std::sort(results.begin(), results.end(), DtaResult::compareSize);
+		} else if (optOG) {
+			// Directories first, then files
+			std::sort(results.begin(), results.end(), DtaResult::groupDirs);
+		} else if (optO) {
+			// Directories first, then files, both sort by name
+			std::sort(results.begin(), results.end(), DtaResult::groupDef);
+		}
```

Comment:
(maybe?)

``` c++
		DtaResult sort_by == DtaResult::None;
		if (optON)
			sort_by = DtaResult::compareName;
		else if (optOE)
			sort_by = DtaResult::compareExt;
		else if (optOD)
			sort_by = DtaResult::compareDate);
		else if (optOS)
			sort_by = DtaResult::compareSize);
		else if (optOG) // Directories first, then files
			sort_by = DtaResult::groupDirs);
		else if (optO) // Directories first, then files, both sort by name
			sort_by = DtaResult::groupDef);

		(if sort_by != DtaResult::None)
			std::sort(results.begin(), results.end(), sort_by);
```

Reply:
The current code in this PR cannot be replaced by the new code you suggested as is (apparently it requires some work to get it compile). Maybe just keep it as is for now, and possibly a cleanup PR later.